cloudevents/formats/avro-turbo-format.md

```diff
@@ -0,0 +1,217 @@
+# Avro Turbo Event Format for CloudEvents - Version 1.0.3-wip
+
+## Abstract
+
+The Avro Turbo Format for CloudEvents defines how events attributes are expressed in
+the [Avro 1.9.0 Specification][avro-spec].
+
+This differs from the [Avro format](avro-format.md) in that:
+
+- It is optimized for performance, preferring a more compact representation.
+- Only supports spec version 1.0 (any changes to spec version, requires 
+  changes to Avro schema, which changes the finger-print, 
+  breaking compatibility).
+- Does not natively support JSON (JSON can be straight-forwardly serialized 
+  to bytes and this was therefore not considered neccessary).
+
+## Table of Contents
+
+1. [Introduction](#1-introduction)
+2. [Attributes](#2-attributes)
+3. [Data](#3-data)
+4. [Transport](#4-transport)
+5. [Examples](#5-examples)
+
+## 1. Introduction
+
+[CloudEvents][ce] is a standardized and protocol-agnostic definition of the
+structure and metadata description of events. This specification defines how the
+elements defined in the CloudEvents specification are to be represented in the
+[Avro 1.9.0][avro-primitives].
+
+The [Attributes](#2-attributes) section describes the naming conventions and
+data type mappings for CloudEvents attributes for use as Avro message
+properties.
+
+This specification does not define an envelope format. The Avro type system's
+intent is primarily to provide a consistent type system for Avro itself and not
+for message payloads.
+
+The Avro event format does not currently define a batch mode format.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+## 2. Attributes
+
+This section defines how CloudEvents attributes are mapped to the Avro
+type-system. This specification explicitly maps each attribute.
+
+### 2.1 Type System Mapping
+
+The CloudEvents type system MUST be mapped to Avro types as follows.
+
+| CloudEvents   | Avro                                                                   |
+| ------------- | ---------------------------------------------------------------------- |
+| Boolean       | [boolean][avro-primitives]                                             |
+| Integer       | [int][avro-primitives]                                                 |
+| String        | [string][avro-primitives]                                              |
+| Binary        | [bytes][avro-primitives]                                               |
+| URI           | [string][avro-primitives] following [RFC 3986 §4.3][rfc3986-section43] |
+| URI-reference | [string][avro-primitives] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [long][avro-primitives]                                                |
```

Comment:
I'd note here that this restricts the precision to milliseconds. (Assuming I've read things correctly later.) We also need to say somewhere that this is the number of milliseconds since 1970-01-01T00:00:00Z. (I'm okay with handwaving between UTC and TAI...)

Reply:
> The smallest time unit that RFC3339 can support is one nanosecond (1 ns).

Reply:
In practical cases, milli seconds will be enough. However, IHMO we should not truncate time to save space.

Reply:
> a nanosecond is smaller than a microsecond.

> 1 microsecond (µs) = 1000 nanoseconds (ns)

Reply:
> A long cannot represent a nanosecond.